components/sup/src/manager/mod.rs

```diff
@@ -934,138 +1096,213 @@ impl Manager {
+            Arc::clone(&self.updater),
+            Arc::clone(&self.busy_services),
+            Arc::clone(&self.reconcile_services),
+        )
+    }
 
-        // The problem we're trying to work around here by adding this block is that `write`
-        // creates an immutable borrow on `self`, and `self.remove_service` needs `&mut self`.
-        // The solution is to introduce the block to drop the immutable borrow before the call to
-        // `self.remove_service`, and use `mem::swap` to move the services to a variable defined
-        // outside the block while we have the lock.
-        {
-            let mut services = self
-                .state
-                .services
+    /// Remove the given service from the manager.
+    fn service_stop_future(
+        service: Service,
+        user_config_watcher: Arc<RwLock<UserConfigWatcher>>,
+        updater: Arc<Mutex<ServiceUpdater>>,
+        busy_services: Arc<Mutex<HashSet<PackageIdent>>>,
+        reconcile_services: Arc<AtomicBool>,
+    ) -> impl Future<Item = (), Error = ()> {
+        // JW TODO: Update service rumor to remove service from
+        // cluster
+        // TODO (CM): But only if we're not going down for a restart.
+        let ident = service.spec_ident.clone();
+        let stop_it = service.stop().then(move |_| {
+            user_config_watcher
                 .write()
-                .expect("Services lock is poisoned!");
-            mem::swap(services.deref_mut(), &mut svcs);
-        }
+                .expect("Watcher lock poisoned")
+                .remove(&service);
+            updater
+                .lock()
+                .expect("Updater lock poisoned")
+                .remove(&service);
+            Ok(())
+        });
+        Self::wrap_async_service_operation(ident, busy_services, reconcile_services, stop_it)
+    }
+
+    /// Wrap a future that starts, stops, or restarts a service with
+    /// logic that marks that service as "busy" for the duration of
+    /// the process.
+    ///
+    /// This allows us to postpone taking additional action on
+    /// services until they're done with what they're doing.  (For
+    /// example, consider stopping a service that takes 10 seconds to
+    /// shut down (including post-stop hook execution), but then
+    /// executing `hab svc start SERVICE` 2 seconds into that 10
+    /// seconds.)
+    ///
+    /// As more service operations (e.g., hooks) become asynchronous,
+    /// we'll need to wrap those operations in this logic to ensure
+    /// consistent operation.
+    fn wrap_async_service_operation<F>(
+        ident: PackageIdent,
+        busy_services: Arc<Mutex<HashSet<PackageIdent>>>,
+        reconcile_services: Arc<AtomicBool>,
+        fut: F,
+    ) -> impl Future<Item = (), Error = ()>
+    where
+        F: IntoFuture<Item = (), Error = ()>,
+    {
+        // TODO (CM): can't wait for the Pinning API :(
+        let busy_services_2 = busy_services.clone();
```

Comment:
It _is_ technically possible to pass things through, but yes, that would end up changing the type of the future. That would also couple whatever future we passed in with internal details of the manager, which seems fragile and makes things subject to unnecessary churn as we modify this code.

As such, I'd rather leave the code as-is for now. The cloning is just of an `Arc`, so that's pretty cheap. And once the pinning API is in place, I think all that cloning goes away entirely.

I'll change to using `Arc::clone`, though 👍